flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/SinkReuser.java

```diff
@@ -0,0 +1,213 @@
+        List<ReusableSinkGroup> reusableSinkGroups = groupReusableSink(allSinkNodes);
+
+        Set<Sink> reusedSinkNodes = reuseSinkAndAddUnion(reusableSinkGroups);
+
+        // Remove all unused sink nodes
+        return relNodes.stream()
+                .filter(root -> !(root instanceof Sink) || reusedSinkNodes.contains(root))
+                .collect(Collectors.toList());
+    }
+
+    private Set<Sink> reuseSinkAndAddUnion(List<ReusableSinkGroup> reusableSinkGroups) {
+        final Set<Sink> reusedSinkNodes = Collections.newSetFromMap(new IdentityHashMap<>());
+        reusableSinkGroups.forEach(
+                group -> {
+                    List<Sink> originalSinks = group.originalSinks;
+                    if (originalSinks.size() <= 1) {
+                        Preconditions.checkState(originalSinks.size() == 1);
+                        reusedSinkNodes.add(originalSinks.get(0));
+                        return;
+                    }
+                    List<RelNode> allSinkInputs = new ArrayList<>();
+                    for (Sink sinkNode : originalSinks) {
+                        allSinkInputs.add(sinkNode.getInput());
+                    }
+                    // Use the first sink node as the final reused sink node
+                    Sink reusedSink = originalSinks.get(0);
+
+                    Union unionForReusedSinks;
+
+                    unionForReusedSinks =
+                            new BatchPhysicalUnion(
+                                    reusedSink.getCluster(),
+                                    group.inputTraitSet,
+                                    allSinkInputs,
+                                    true,
+                                    // use sink input row type
+                                    reusedSink.getRowType());
+
+                    reusedSink.replaceInput(0, unionForReusedSinks);
+                    reusedSinkNodes.add(reusedSink);
+                });
+        return reusedSinkNodes;
+    }
+
+    /**
+     * Grouping sinks that can be reused with each other.
+     *
+     * @param allSinkNodes in the plan.
+     * @return a list contains all grouped sink.
+     */
+    private List<ReusableSinkGroup> groupReusableSink(List<Sink> allSinkNodes) {
+        List<ReusableSinkGroup> reusableSinkGroups = new ArrayList<>();
+
+        for (Sink currentSinkNode : allSinkNodes) {
+            boolean canBeReused = false;
+            String currentSinkDigest = getDigest(currentSinkNode);
+            SinkAbilitySpec[] currentSinkSpecs;
+
+            currentSinkSpecs = ((BatchPhysicalSink) currentSinkNode).abilitySpecs();
+            RelTraitSet currentInputTraitSet = currentSinkNode.getInput().getTraitSet();
+            for (ReusableSinkGroup group : reusableSinkGroups) {
+                // Only table sink with the same digest, specs and input trait set can be reused
+                if (!(group.digest.equals(currentSinkDigest)
```

Comment:
it would be nicer code to have the if as a positive if and then have an else to avoid the continue.

Reply:
Revised